rust-client/src/main.rs

```diff
@@ -54,233 +55,378 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
-        drive_hole_punch(swarm, remote_peer_id).await
-    };
 
-    let finish_time = SystemTime::now();
-    let elapsed_time = (finish_time.duration_since(UNIX_EPOCH).unwrap()
-        - start_time.duration_since(UNIX_EPOCH).unwrap())
-    .as_secs_f32();
-    let request = tonic::Request::new(grpc::TrackHolePunchRequest {
-        client_id: local_peer_id.to_bytes(),
-        remote_id: remote_peer_id.to_bytes(),
-        success: result.is_ok(),
-        started_at: start_time
-            .duration_since(UNIX_EPOCH)
-            .unwrap()
-            .as_millis()
-            .try_into()
-            .unwrap(),
-        remote_multi_addresses: remote_addrs.into_iter().map(|a| a.to_vec()).collect(),
-        // TODO: Set proper attempts
-        attempts: 0,
-        // TODO
-        error: String::new(),
-        // TODO
-        direct_dial_error: String::new(),
-        // TODO
-        start_rtt: 0.0,
-        elapsed_time,
-        end_reason: grpc::HolePunchEndReason::Unknown.into(),
-    });
-
-    client.track_hole_punch(request).await?;
+        let request = HolePunchState::new(local_peer_id, remote_peer_id, remote_addrs)
+            .drive_hole_punch(&mut swarm)
+            .await;
+
+        client
+            .track_hole_punch(tonic::Request::new(request))
+            .await?;
+
+        futures_timer::Delay::new(std::time::Duration::from_secs(1)).await;
+    }
 
     Ok(())
 }
 
-fn build_swarm(local_key: identity::Keypair) -> Swarm<Behaviour> {
+async fn build_swarm(
+    local_key: identity::Keypair,
+) -> Result<Swarm<Behaviour>, Box<dyn std::error::Error>> {
     let local_peer_id = PeerId::from(local_key.public());
-    let (relay_transport, relay_client) = Client::new_transport_and_behaviour(local_peer_id);
+    let transport = DnsConfig::system(TcpConfig::new().port_reuse(true)).await?;
+    let (relay_transport, relay_behaviour) =
+        new_transport_and_behaviour(RelayConfig::default(), transport);
 
     let noise_keys = noise::Keypair::<noise::X25519Spec>::new()
         .into_authentic(&local_key)
         .expect("Signing libp2p-noise static DH keypair failed.");
 
-    let transport = OrTransport::new(
-        relay_transport,
-        block_on(DnsConfig::system(TcpConfig::new().port_reuse(true))).unwrap(),
-    )
-    .upgrade(upgrade::Version::V1)
-    .authenticate(noise::NoiseConfig::xx(noise_keys).into_authenticated())
-    // TODO: Consider supporting mplex.
-    .multiplex(libp2p::yamux::YamuxConfig::default())
-    .boxed();
-
-    let identify_config = IdentifyConfig::new("/TODO/0.0.1".to_string(), local_key.public())
+    let transport = relay_transport
+        .upgrade(upgrade::Version::V1)
+        .authenticate(noise::NoiseConfig::xx(noise_keys).into_authenticated())
+        // TODO: Consider supporting mplex.
+        .multiplex(libp2p::yamux::YamuxConfig::default())
+        .boxed();
+
+    let identify_config = IdentifyConfig::new("/identify/0.0.1".to_string(), local_key.public())
```

Comment:
Ah, this is confusing. The first parameter is the `protocol_version`, though what is meant is not the identify protocol version, but the protocol version of the whole p2p network:

https://github.com/libp2p/specs/tree/master/identify#protocolversion

Not sure what to use here. Likely just what every IPFS node uses, as we are connecting to the IPFS network.